pkg/resource/labels.go

```diff
@@ -0,0 +1,109 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package resource
+
+import "fmt"
+
+// Labels is a set free-form of key-value pairs.
+//
+// Order of keys is not guaranteed.
+type Labels struct {
+	m map[string]string
+}
+
+// Set the label.
+//
+// Setting the label copies the map, so resources share common labels as long as possible.
+func (labels *Labels) Set(key, value string) {
+	if labels.m == nil {
+		labels.m = make(map[string]string)
+	} else {
+		v, ok := labels.m[key]
+		if ok && v == value {
+			// no change
+			return
+		}
+
+		labelsCopy := make(map[string]string, len(labels.m))
+
+		for k, v := range labels.m {
+			labelsCopy[k] = v
+		}
+
+		labels.m = labelsCopy
+	}
+
+	labels.m[key] = value
+}
+
+// Get the label.
+func (labels *Labels) Get(key string) (string, bool) {
+	if labels.m == nil {
+		return "", false
+	}
+
+	value, ok := labels.m[key]
+
+	return value, ok
+}
+
+// Raw returns the labels map.
+//
+// Label map should not be modified outside of the call.
+func (labels *Labels) Raw() map[string]string {
+	return labels.m
+}
+
+// Equal checks label for equality.
+func (labels Labels) Equal(other Labels) bool {
+	if labels.m == nil && other.m == nil {
```

Comment:
Do we differentiate between nil map and empty map?

Reply:
that's a good question; also there's no way to delete labels right now, so there's no empty map ever.

Reply:
but I think we should handle that

Reply:
added the handling and a delete method